### PR TITLE
skaffold init and buildpacks: skip dependencies

### LINKED ISSUE #3568: skaffold init looks for go.mod files under vendor/
### Expected behavior

```
$ skaffold init --analyze --XXenableBuildpacksInit | jq .
{
  "builders": [
    {
      "name": "Buildpacks",
      "payload": {
        "path": "go.mod"
      }
    }
  ],
  "images": [
    {
      "name": "github.com/google/go-containerregistry/pkg/authn/k8schain/tests/explicit",
      "foundMatch": false
    },
    {
      "name": "github.com/google/go-containerregistry/pkg/authn/k8schain/tests/implicit",
      "foundMatch": false
    },
    {
      "name": "github.com/google/go-containerregistry/pkg/authn/k8schain/tests/noauth",
      "foundMatch": false
    },
    {
      "name": "github.com/google/go-containerregistry/pkg/authn/k8schain/tests/serviceaccount",
      "foundMatch": false
    }
  ]
}
```

### Actual behavior

```
$ skaffold init --analyze --XXenableBuildpacksInit | jq .
{
  "builders": [
    {
      "name": "Buildpacks",
      "payload": {
        "path": "go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/Azure/go-autorest/autorest/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/Azure/go-autorest/autorest/adal/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/Azure/go-autorest/autorest/date/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/Azure/go-autorest/autorest/to/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/Azure/go-autorest/autorest/validation/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/Azure/go-autorest/logger/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/Azure/go-autorest/tracing/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/Microsoft/go-winio/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/google/gofuzz/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/json-iterator/go/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/konsorten/go-windows-terminal-sequences/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/maxbrunsfeld/counterfeiter/v6/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/russross/blackfriday/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/sirupsen/logrus/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/spf13/cobra/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/spf13/pflag/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/github.com/vdemeester/k8s-pkg-credentialprovider/go.mod"
      }
    },
    {
      "name": "Docker",
      "payload": {
        "path": "vendor/golang.org/x/net/http2/Dockerfile"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/golang.org/x/oauth2/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/google.golang.org/appengine/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/gopkg.in/yaml.v2/go.mod"
      }
    },
    {
      "name": "Buildpacks",
      "payload": {
        "path": "vendor/k8s.io/klog/go.mod"
      }
    }
  ],
  "images": [
    {
      "name": "github.com/google/go-containerregistry/pkg/authn/k8schain/tests/explicit",
      "foundMatch": false
    },
    {
      "name": "github.com/google/go-containerregistry/pkg/authn/k8schain/tests/implicit",
      "foundMatch": false
    },
    {
      "name": "github.com/google/go-containerregistry/pkg/authn/k8schain/tests/noauth",
      "foundMatch": false
    },
    {
      "name": "github.com/google/go-containerregistry/pkg/authn/k8schain/tests/serviceaccount",
      "foundMatch": false
    }
  ]
}
```
### Information

- Skaffold version: 4c38a79c3a59a14cf4cae14015bbcfdf9816c491
- Operating system: Linux

### Steps to reproduce the behavior

1. cd https://github.com/google/go-containerregistry
2. `skaffold init --analyze --XXenableBuildpacksInit`
3. use eyeballs

### PR DESCRIPTION
**Pull request by dgageot**
Fixes #3568

Signed-off-by: David Gageot <david@gageot.net>

### CODE REVIEW

File: pkg/skaffold/build/buildpacks/init.go [hunk 1]
```diff
@@ -69,7 +71,22 @@ func (c ArtifactConfig) Path() string {
 
 // validate checks if a file is a valid Buildpack configuration.
 func validate(path string) bool {
-	name := filepath.Base(path)
+	switch filepath.Base(path) {
+	case "package.json":
+		return !hasParent(path, "node_modules")
+	case "go.mod":
+		return !hasParent(path, "vendor")
```
Comment (briandealwis) [inline pkg/skaffold/build/buildpacks/init.go]: Seems worth emitting logging something here to inform the user that the project was skipped due to node_modules/vendor.  I know I'd puzzled as to why one "clean" project is picked up both another ignored.  I'm sure there are a number of projects that commit their vendor'd folders (like skaffold).

And a todo to correlate this with the [build.exclude](https://github.com/buildpacks/rfcs/blob/master/text/0019-project-descriptor.md#buildinclude-and-buildexclude)?

Comment (dgageot) [reply pkg/skaffold/build/buildpacks/init.go]: I'm not sure I understand your point. We are ignoring what's under `vendor` and `node_modules`. Wether those folders exist or not give the same result.

File: pkg/skaffold/build/buildpacks/init_test.go [hunk 2]
```diff
@@ -32,17 +32,27 @@ func TestValidate(t *testing.T) {
 	}{
 		{
 			description:   "NodeJS",
-			path:          "path/to/package.json",
+			path:          filepath.Join("path", "to", "package.json"),
```
Comment (tejal29) [inline pkg/skaffold/build/buildpacks/init_test.go]: just for completeness add corner test cases for
`go.mod`, `package.json` -> i.e. no filepath.Base = empty string
and `somedir/vendor/go.mod` -> i.e vendor not first parent.

File: pkg/skaffold/build/buildpacks/init.go (hunk 1 above)
Comment (briandealwis) [reply pkg/skaffold/build/buildpacks/init.go]: :blush: I'm sorry, I did not dig into this code hard enough.  From our previous conversations about interference with buildpacks from projects with `node_modules` or `vendor`, the name `validate` threw me off: I thought this code was to identify problematic projects with an accompanying `node_modules` or `vendor`.

File: pkg/skaffold/build/buildpacks/init.go [hunk 3]
```diff
@@ -69,7 +71,22 @@ func (c ArtifactConfig) Path() string {
 
 // validate checks if a file is a valid Buildpack configuration.
 func validate(path string) bool {
-	name := filepath.Base(path)
+	switch filepath.Base(path) {
+	case "package.json":
+		return !hasParent(path, "node_modules")
```
Comment (balopat) [inline pkg/skaffold/build/buildpacks/init.go]: how about using [filepath.Match](https://golang.org/pkg/path/filepath/#Match) instead?

**Pull request merged**